### PR TITLE
Add ping/pong to websocket protocol

### LINKED ISSUE #828: Add ping/pong protocol to control websocket
Reverse proxies like Nginx will close connections (including websockets) after certain application-level inactivity (i.e. tcp keepalive is not good enough). 

In order to avoid this we should add a ping/pong keepalive protocol to the control websocket.

References:

* [Websockets in Ngninx](https://blog.martinfjordvald.com/2013/02/websockets-in-nginx/):

> WebSockets are still affected by proxy_read_timeout which defaults to 60 seconds. This means that if you have an application using WebSockets but not sending any data more than once per 60 seconds you either need to increase the timeout or implement a ping message to keep the connection alive. 

* https://www.reddit.com/r/golang/comments/3bt1ej/gorilla_websockets_start_closing_after_20_clients/cspc6st

### PR DESCRIPTION
**Pull request by paulbellamy**
Fixes #828

Also made the websocket codec `WaitForReadError` return the error.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Also needed for the pipes websocket, no?

And do we ever call codec.Close()?

**Comment by paulbellamy** [comment]
re: pipes. probably, actually.

`codec.Close()` is called from the rpc client here: https://golang.org/src/net/rpc/client.go?s=7387:7422#L267

**Comment by paulbellamy** [comment]
@tomwilkie, added ping to all probe/app websockets (I think), wdyt?

File: common/xfer/websocket.go [hunk 1]
```diff
@@ -2,18 +2,110 @@ package xfer
...
+type pingingWebsocket struct {
+	pinger    *time.Ticker
+	quit      chan struct{}
+	readLock  sync.Mutex
+	writeLock sync.Mutex
+	*websocket.Conn
+}
+
+// Ping adds a periodic ping to a websocket connection.
+func Ping(c *websocket.Conn) Websocket {
```
Comment (paulbellamy) [inline common/xfer/websocket.go]: Probably want a better name for this...

**Comment by paulbellamy** [comment]
Tested this by opening a websocket (via nginx) to /api/control/ws, and seeing it be terminated after 60s. With this patch (and `xfer.Ping` added on the server-side instead of the client-side), it stays open indefinitely.

File: app/api_topology.go [hunk 2]
```diff
@@ -79,17 +78,18 @@ func handleWebsocket(
 	renderer render.Renderer,
 	loop time.Duration,
 ) {
-	conn, err := upgrader.Upgrade(w, r, nil)
+	wsConn, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
 		// log.Info("Upgrade:", err)
 		return
 	}
+	conn := xfer.Ping(wsConn)
```
Comment (tomwilkie) [inline app/api_topology.go]: Can we provide an xfer.Upgrade which automatically makes a pinger etc?  So you don't have to do this two-step everywhere.

File: common/xfer/websocket.go [hunk 3]
```diff
@@ -2,29 +2,104 @@ package xfer
...
+	p.pinger = time.AfterFunc(pingPeriod, p.ping)
+	p.ping()
+	return p
+}
+
+func (p *pingingWebsocket) ping() {
+	p.writeLock.Lock()
+	defer p.writeLock.Unlock()
+	if err := p.conn.WriteControl(websocket.PingMessage, nil, mtime.Now().Add(writeWait)); err != nil {
+		p.Close()
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: Should the err be logged at all?

File: common/xfer/websocket.go [hunk 4]
```diff
@@ -2,29 +2,104 @@ package xfer
...
+}
+
+// ReadMessage is a helper method for getting a reader using NextReader and
+// reading from that reader to a buffer.
+func (p *pingingWebsocket) ReadMessage() (int, []byte, error) {
+	p.readLock.Lock()
+	defer p.readLock.Unlock()
+	messageType, b, err := p.conn.ReadMessage()
+	if err == nil {
+		err = p.conn.SetReadDeadline(mtime.Now().Add(pongWait))
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: Why is the read deadline set after doing the read?

File: common/xfer/websocket.go [hunk 5]
```diff
@@ -2,29 +2,104 @@ package xfer
...
+// WriteJSON writes the JSON encoding of v to the connection.
+func (p *pingingWebsocket) WriteJSON(v interface{}) error {
+	p.writeLock.Lock()
+	defer p.writeLock.Unlock()
+	w, err := p.conn.NextWriter(websocket.TextMessage)
 	if err != nil {
 		return err
 	}
 	err1 := codec.NewEncoder(w, &codec.JsonHandle{}).Encode(v)
+	p.conn.SetWriteDeadline(mtime.Now().Add(writeWait))
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: Shouldn't this be set before doing the write?

File: common/xfer/websocket.go [hunk 6]
```diff
@@ -44,5 +121,25 @@ func ReadJSONfromWS(c *websocket.Conn, v interface{}) error {
 		// One value is expected in the message.
 		err = io.ErrUnexpectedEOF
 	}
+	if err == nil {
+		p.conn.SetReadDeadline(mtime.Now().Add(pongWait))
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: Ditto

File: probe/appclient/app_client.go [hunk 7]
```diff
@@ -188,13 +188,12 @@ func (c *appClient) controlConnection() (bool, error) {
...
 	url := sanitize.URL("ws://", 0, "/api/control/ws")(c.target)
-	conn, _, err := c.wsDialer.Dial(url, headers)
+	wsConn, _, err := c.wsDialer.Dial(url, headers)
 	if err != nil {
 		return false, err
 	}
-	defer func() {
-		conn.Close()
-	}()
+	conn := xfer.Ping(wsConn)
```
Comment (tomwilkie) [inline probe/appclient/app_client.go]: Similarly, can we have an xfer.DialWS?

File: probe/appclient/app_client.go [hunk 8]
```diff
@@ -280,6 +279,7 @@ func (c *appClient) pipeConnection(id string, pipe xfer.Pipe) (bool, error) {
 	if err != nil {
 		return false, err
 	}
+	conn := xfer.Ping(wsConn)
```
Comment (tomwilkie) [inline probe/appclient/app_client.go]: Ditto.

**Comment by tomwilkie** [comment]
Looking better.  If I understand this right, the probe will now ping the app and the app ping the probe?  Not problem with that I guess.

File: common/xfer/websocket.go (hunk 4 above)
Comment (paulbellamy) [inline common/xfer/websocket.go]: It's the next time we need to see something from the other side before disconnecting.

File: common/xfer/websocket.go (hunk 5 above)
Comment (paulbellamy) [inline common/xfer/websocket.go]: ah, probably. Was thinking the `Close()` is the write.

**Comment by tomwilkie** [comment]
```
./common/xfer/websocket.go:67:1: comment on exported function DialWS should be of the form "DialWS ..."
```

File: common/xfer/websocket.go [hunk 9]
```diff
@@ -44,5 +156,25 @@ func ReadJSONfromWS(c *websocket.Conn, v interface{}) error {
 		// One value is expected in the message.
 		err = io.ErrUnexpectedEOF
 	}
+	if err == nil {
+		err = p.conn.SetReadDeadline(mtime.Now().Add(pongWait))
+	}
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: I think we decided to remove this, and rely on pong.

File: common/xfer/websocket.go (hunk 4 above)
Comment (tomwilkie) [inline common/xfer/websocket.go]: decided to remove and rely on pong

File: common/xfer/websocket.go [hunk 10]
```diff
@@ -2,28 +2,138 @@ package xfer
...
+	if err != nil {
+		return nil, nil, err
+	}
+	return Ping(wsConn), resp, nil
+}
+
+// Ping adds a periodic ping to a websocket connection.
+func Ping(c *websocket.Conn) Websocket {
+	p := &pingingWebsocket{conn: c}
+	p.conn.SetPongHandler(p.pong)
```
Comment (tomwilkie) [inline common/xfer/websocket.go]: set initial read timeout.

**Comment by tomwilkie** [comment]
LGTM once tests pass.

**Pull request merged**